addons/fleet/fleet.py

```diff
@@ -1,818 +1,731 @@
-        'odometer': fields.function(_get_odometer, fnct_inv=_set_odometer, type='float', string='Odometer Value', help='Odometer measure of the vehicle at the moment of this log'),
-        'odometer_unit': fields.related('vehicle_id', 'odometer_unit', type="char", string="Unit", readonly=True),
-        'date' :fields.date('Date',help='Date when the cost has been executed'),
-        'contract_id': fields.many2one('fleet.vehicle.log.contract', 'Contract', help='Contract attached to this cost'),
-        'auto_generated': fields.boolean('Automatically Generated', readonly=True, required=True),
-    }
+        odometer = self.env['fleet.vehicle.odometer'].create({
+            'value': self.odometer,
+            'date': self.date or fields.Date.today(),
+            'vehicle_id': self.vehicle_id.id})
+        self.odometer_id = odometer.id
+
+    name = fields.Char(related='vehicle_id.name', store=True)
+    vehicle_id = fields.Many2one('fleet.vehicle', string='Vehicle', required=True,
+                                 help='Vehicle concerned by this log')
+    cost_subtype_id = fields.Many2one('fleet.service.type', string='Type',
+                                      help='Cost type purchased with this cost')
+    amount = fields.Float(string='Total Price')
+    cost_type = fields.Selection(selection=[('contract', 'Contract'), ('services', 'Services'),
+                                            ('fuel', 'Fuel'), ('other', 'Other')],
+                                 string='Category of the cost', help='For internal purpose only',
+                                 required=True, default='other')
+    parent_id = fields.Many2one('fleet.vehicle.cost', string='Parent', help='Parent cost to this current cost')
+    cost_ids = fields.One2many('fleet.vehicle.cost', 'parent_id', string='Included Services')
+    odometer_id = fields.Many2one('fleet.vehicle.odometer', string='Odometer',
+                                  help='Odometer measure of the vehicle at the moment of this log')
+    odometer = fields.Float(compute='_compute_get_odometer', inverse='_compute_set_odometer', string='Odometer Value',
+                            help='Odometer measure of the vehicle at the moment of this log', store=True)
+    odometer_unit = fields.Selection(related='vehicle_id.odometer_unit', string="Unit", readonly=True)
+    date = fields.Date(help='Date when the cost has been executed')
+    contract_id = fields.Many2one('fleet.vehicle.log.contract', string='Contract',
+                                  help='Contract attached to this cost')
+    auto_generated = fields.Boolean('Automatically Generated', readonly=True, required=True)
+
+    @api.model
+    def create(self, values):
+        # make sure that the data are consistent with values of parent and contract records given
+        if 'parent_id' in values and values['parent_id']:
```

Comment:
@jas-odoo then what will be output of self.browse(values['parent_id']),  it's many2one field